src/lib.rs

```diff
@@ -111,24 +103,18 @@ where
     where
         P: GetNextMut<NextPtr = Option<P>>,
     {
-        unsafe {
-            let next = value.get_next() as *mut Option<P>;
-            let raw = value.into_raw();
-            // Iff next was set to Some(P) we want to
-            // assert that it was droppeds
-            drop(ptr::read(next));
-            loop {
-                let pcurrent = self.inner.load(load_order);
-                let current = if pcurrent.is_null() {
-                    None
-                } else {
-                    Some(FromRawPtr::from_raw(pcurrent))
-                };
-                ptr::write(next, current);
-                let last = self.inner.compare_and_swap(pcurrent, raw, cas_order);
-                if last == pcurrent {
-                    return last.is_null();
-                }
+        let next = value.get_next() as *mut Option<P>;
+        let raw = value.into_raw();
+        // Iff next was set to Some(P) we want to
+        // assert that it was droppeds
+        drop(unsafe { ptr::read(next) });
```

Comment:
It's not a new bug, but imagine what this code does for `Arc`. It's completely incorrect.

Reply:
No, it's correct; the new head of the list is going to have its next changed to point to the current head, thus it's next value needs to be dropped (normally it would just be None anyway).

Reply:
Yeah, realized that earlier but forgot to comment here. It should use `drop_in_place` though (or maybe a mem swap works, too).

Reply:
Yeah, could be drop_in_place. There's not much point in a swap/replace since you don't want to drop any of the later values in the loop.

Reply:
I meant `mem::replace`, you could just write the new pointer into it I think, but didn't look into it yet.

Reply:
Not sure I understand your comment.

Reply:
The first time around the loop you could mem::replace in the new pointer. But the second time around the loop you can't drop the old value, so you'd have to ptr::write, so you'd basically have to just unroll the loop once, and that seems pointless.

Reply:
Ah right it has to retry until the snapshot matches, should probably read the code before I comment rubbish here, sorry.

Reply:
@dbkaplun Could we merge PR for PR? I'm kind of confused by which PR depends on what exactly. I suppose merging #4 first is what you intended?

Reply:
@torkleyy yes if we could merge #4 that would be great. It shouldn't have any substantive changes other than reducing scope of `unsafe` where possible.

Reply:
Could you resolve the comments I made first?